docs/man/tcdemux.1
.TH tcdemux 1 "09th January 2003" "tcdemux(1)"
.SH NAME
tcdemux \- demultiplex a program stream
.SH SYNOPSIS
.TP
.B tcdemux
.B -i
.I name
[
.B -t
.I magic
] [
.B -x
.I codec
] [
.B -S
.I unit,[s1-s2]
] [
.B -a
.I ach,[vch]
] [
.B -s
.I 0xnn
] [
.B -M
.I mode
] [
.B -f
.I fps
] [
.B -W
] [
.B -O
] [
.B -P
.I name
] [
.B -A
.I n[,m[,...]]
] [
.B -d
.I verbosity
] [
.B -v
]
.SH COPYRIGHT
\fBtcdemux\fP is Copyright (C) by Thomas Oestreich.
.SH DESCRIPTION
.B tcdemux
is part of and usually called by \fBtranscode\fP.
.br
However, it can also be used independently.
.br
\fBtcdemux\fP reads source (from stdin if not explicitely defined) and
prints on the standard output.
.SH OPTIONS
.IP "\fB-i \fIname\fP"
Specify input source.  If ommited, \fIstdin\fP is assumed.
.br
.IP "\fB-x \fIcodec\fP"
Process only packets containing given
.I codec
as payload.
Examples are "mpeg2" for MPEG video or "ac3" for audio packets.
.br
.IP "\fB-t \fImagic\fP"
Specify the input file type
.I magic.
Useful for using
.B tcdemux
in pipelines. This is normally autodetected.
.br
.IP "\fB-a \fIach[,vch]\fP"
Process only packets containing given audio track
.I ach
and video track
.I vch.
Most program stream feature multiple audio tracks but only a single video stream. Default is "-a 0,0".
.br
.IP "\fB-s \fI0xnn\fP"
Select specific multimedia track using the standard identifier
.I 0xnn.
This useful to extract subtitle packets.
.br
.IP "\fB-f \fIfps\fP"
Specify the frame rate
.I fps
for subtitle meta information used by external applications.
.br
.IP "\fB-S \fIunit[,s1-s2]\fP"
Select a program unit
.I unit
and a selected GOP range
.I s1-s2
for processing. A
.I unit
is usually a set of GOPs that need to be synchronized
at the very beginning. This option can be used to eleminate junk units at the beginning of the main presentation or to select certain episodes of a TV series from a DVD. A GOP is a logical unit to be properly decoded, i.e., decoding can only start at the beginning of a GOP.
.br
.IP "\fB-O"
Do not skip initial sequence. Used internally by transcode.
.br
.IP "\fB-A \fIn[,m[,...]]\fP"
Select packets using an identifier for extracting only selected streams without processing. This is useful for size reduction of your multimedia stream. Example:

.B tcdemux -i big_dvd.vob -A 0xe0,0x81,0x20 > small_dvd.vob

extracts all packets for the video stream, AC3 audio track (1) and the
first subtitle stream (0).
.br
.IP "\fB-M \fImode\fP"
Select synchronization strategy
.I mode.
This is how transcode selects the beginning of a video and audio stream to
assure both streams are sychronized for further processing:

0	no synchronization. Packets are printed directly to \fIstdout\fP.

1	synchronization based on PTS information found in the packet header. Audio packets are delayed until a suitable starting point is found. Default for PAL videos.

2	NTSC adapted synchronization mode. Detailed information for each GOP contained in the video stream are piped to a \fIstdout\fP to be processed by
the certain import modules. This enables adjusting the frame rate or inverse telecine, if necessary.

3       more advanced synchronization mode 1 with msec granular adjustment. Details delegated to transcode.

4       more advanced synchronization mode 2 with msec granular adjustment for NTSC materail. Details delegated to transcode.

5-7	undocumented debugging modes for internal use.
.br
.IP "\fB-P \fIfile\fP"
Write GOP meta information to
.I file
instead of \fstdout\fP.
.br
.IP \fB-W
Print a navigation log file for a given video stream to \fIstdout\fP. This is used for transcode's "psu mode" and "cluster mode".
.br
.IP "\fB-d\fP \fIlevel\fP"
With this option you can specify a bitmask to enable different levels
of verbosity (if supported).  You can combine several levels by adding the
corresponding values:

QUIET         0

INFO          1

DEBUG         2

STATS         4

WATCH         8

FLIST        16

VIDCORE      32

SYNC         64

COUNTER     128

PRIVATE     256
.IP "\fB-v\fP"
Print version information and exit.
.SH NOTES
\fBtcdemux\fP is a front end for de-multiplexing program streams and is used in \fBtranscode\fP's import modules.
.SH EXAMPLES
The command

.B tcprobe -i movie.vob -x mpeg2

extracts only MPEG video packets. Further processing, i.e., extracting the
elementary video stream is done by
.B tcextract.
.PP
.SH AUTHORS
.B tcdemux
was written by Thomas Oestreich
.br
<ostreich@theorie.physik.uni-goettingen.de> with contributions from
many others.  See AUTHORS for details.
.SH SEE ALSO
.BR avifix (1),
.BR avimerge (1),
.BR avisplit (1),
.BR tcdecode (1),
.BR tcdemux (1),
.BR tcextract (1),
.BR tccat (1),
.BR tcscan (1),
.BR transcode (1)


ChangeLog
Key:
[+] feature added
[!] bugfix of an existing feature
[*] change/improvement (not bugfix) of an existing feature
[-] feature dropped

===========================================================================
transcode-1.2.0
---------------
[+] Brand new export layer with new-style, separate encoders and muxers.
[+] New encoders: theora, vorbis, faac, x264.
[+] New multiplexors: ogg, libavformat, yuv4mpeg2.
[+] New filter: barrel.
[+] ShoutCast support using ogg multiplexor.
[!] Fix crash when using transformations (-Z, -j).
[!] Fix compilation error with x264 version 78.
[*] Improved internal buffering system.
[*] Added basic A/V synchronization module.
[!] Fix misuse of configuration files which leads to crash.
[*] Allow to handle fallback directories for configuration files.
[*] Improved debug messages handling and internal debug API.
[-] The import_imlist is deprecated. Use --multi_input -x im instead.
[*] Big YAIT update by Allan Snider.
[!] Make the -W option work again like 1.0.x branch.
[*] Switch from ImageMagick to GraphicsMagick.
[*] Allow to build without FFmpeg.
[!] Remove outdated libav* information from the FAQ.
[+] Added support for CubicKeys4 and Sinc8 zoom algorithms.
[+] Added module registry to automatically select a module from format
[+] Added `tccfgshow' to inspect the transcode internal capabilities.
[-] Removed the unmantained PVM code.
[+] Added a special-crafted import module targeting webcams using libv4l.
[*] Update the video stabilization plugins to version 0.61.
[*] Added the -X option to tcdecode (let the user specify the acceleration)
[+] Enable versioned, parallel installation.
===========================================================================


import/tcdemux.c
/*
 *  tcdemux.c
 *
 *  Copyright (C) Thomas Oestreich - June 2001
 *
 *  This file is part of transcode, a video stream processing tool
 *
 *  transcode is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *
 *  transcode is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with GNU Make; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include "src/transcode.h"
#include "tccore/tcinfo.h"

#include "libtcutil/xio.h"
#include "ioaux.h"
#include "tc.h"
#include "demuxer.h"

#define EXE "tcdemux"

#define MAX_BUF     1024

int verbose = TC_QUIET;

/* ------------------------------------------------------------
 *
 * auxiliary routines
 *
 * ------------------------------------------------------------*/

void import_exit(int code)
{
    if (verbose & TC_DEBUG)
        tc_log_msg(EXE, "(pid=%d) exit (code %d)", (int) getpid(), code);
    exit(code);
}

/* ------------------------------------------------------------
 *
 * print a usage/version message
 *
 * ------------------------------------------------------------*/

void version(void)
{
    /* print id string to stderr */
    fprintf(stderr, "%s (%s v%s) (C) 2001-2003 Thomas Oestreich,"
                                   " 2003-2010 Transcode Team\n",
                    EXE, PACKAGE, VERSION);
}

static void usage(int status)
{
    version();

    fprintf(stderr,"\nUsage: %s [options]\n", EXE);

    fprintf(stderr,"    -i name          input file name [stdin]\n");
    fprintf(stderr,"    -t magic         input file type [autodetect]\n");
    fprintf(stderr,"    -x codec         process only packs with codec payload [all]\n");
    fprintf(stderr,"    -S unit[,s1-s2]  presentation unit[,s1-s2] sequences [0,all]\n");
    fprintf(stderr,"    -a ach[,vch]     extract audio[,video] track [0,0]\n");
    fprintf(stderr,"    -s 0xnn          sync with private substream id 0xnn [off]\n");
    fprintf(stderr,"    -M mode          demuxer PES A-V sync mode (0=off|1=PTS only|2=full) [1]\n");
    fprintf(stderr,"    -O               do not skip initial sequence\n");
    fprintf(stderr,"    -P name          write synchronization data to file\n");
    fprintf(stderr,"    -W               write navigation data to stdout\n");
    fprintf(stderr,"    -f fps           frame rate [%.3f]\n", PAL_FPS);
    fprintf(stderr,"    -d mode          verbosity mode\n");
    fprintf(stderr,"    -A n[,m[...]]    pass-through packet payload id\n");
    fprintf(stderr,"    -H               sync hard to supplied fps (no smooth drop)\n");
    fprintf(stderr,"    -v               print version\n");

    exit(status);
}


/* ------------------------------------------------------------
 *
 * demuxer thread frontend
 *
 * ------------------------------------------------------------*/

int main(int argc, char *argv[])
{
    info_t ipipe;
    int ch, n, user = 0, demux_mode = TC_DEMUX_SEQ_ADJUST;
    int npass = 0, *pass = NULL, *new_pass = NULL;
    int keep_initial_seq = 0, hard_fps_flag = 0, pack_sl = PACKAGE_ALL;
    int unit_seek = 0, resync_seq1 = 0, resync_seq2 = INT_MAX;
    int a_track = 0, v_track = 0, subid = 0x80;
    double fps = PAL_FPS;
    long stream_stype = TC_STYPE_UNKNOWN;
    long stream_codec = TC_CODEC_UNKNOWN;
    long stream_magic = TC_MAGIC_UNKNOWN;
    long x;
    char *magic = "", *codec = NULL, *name = NULL;
    char *logfile = SYNC_LOGFILE, *str = NULL, *end = NULL;
    //defaults:
    //proper initialization
    memset(&ipipe, 0, sizeof(info_t));

    libtc_init(&argc, &argv);

    while ((ch = getopt(argc, argv, "A:a:d:x:i:vt:S:M:f:P:WHs:O?h")) != -1) {
        switch (ch) {
          case 'i':
            if (optarg[0] == '-') usage(EXIT_FAILURE);
            name = optarg;
            break;

          case 'O':
            keep_initial_seq = 1;
            break;

          case 'P':
            if (optarg[0] == '-') usage(EXIT_FAILURE);
            logfile = optarg;
            break;

          case 'S':
            if (optarg[0] == '-') usage(EXIT_FAILURE);
            n = sscanf(optarg,"%d,%d-%d",
                       &unit_seek, &resync_seq1, &resync_seq2);
            if (n < 0) {
                tc_log_error(EXE, "invalid parameter for option -S");
                usage(EXIT_FAILURE);
            }

            if (unit_seek < 0) {
                tc_log_error(EXE, "invalid unit parameter for option -S");
                usage(EXIT_FAILURE);
            }

            if (resync_seq1 < 0 || resync_seq2 < 0
             || resync_seq1 >= resync_seq2) {
                tc_log_error(EXE, "invalid sequence parameter for option -S");
                usage(EXIT_FAILURE);
            }
            break;

          case 'd':
            if (optarg[0] == '-') usage(EXIT_FAILURE);
            verbose = atoi(optarg);
            break;

          case 'f':
            if (optarg[0] == '-') usage(EXIT_FAILURE);
            fps = atof(optarg);
            break;

          case 'W':
            demux_mode = TC_DEMUX_SEQ_LIST;
            logfile = NULL;
            break;

          case 'H':
            hard_fps_flag = 1;
            break;

          case 'x':
            if (optarg[0] == '-') usage(EXIT_FAILURE);
            codec = optarg;

            if (strcmp(codec,"ac3") == 0) {
                pack_sl = PACKAGE_AUDIO_AC3;
                stream_codec = TC_CODEC_AC3;
            }

            if (strcmp(codec,"mpeg2") == 0) {
                pack_sl = PACKAGE_VIDEO;
                stream_codec = TC_CODEC_MPEG2;
            }

            if (strcmp(codec,"mp3") == 0) {
                pack_sl = PACKAGE_AUDIO_MP3;
                stream_codec = TC_CODEC_MP3;
            }

            if (strcmp(codec,"pcm") == 0) {
                pack_sl = PACKAGE_AUDIO_PCM;
                stream_codec = TC_CODEC_PCM;
            }

            if (strcmp(codec,"ps1") == 0) {
                pack_sl = PACKAGE_SUBTITLE;
                stream_codec = TC_CODEC_SUB;
            }
            break;

          case 't':
            if (optarg[0] == '-') usage(EXIT_FAILURE);
            magic = optarg;
            user = 1;
            break;

          case 's':
            if (optarg[0] == '-') usage(EXIT_FAILURE);
            subid = strtol(optarg, NULL, 16);
            break;

          case 'A':
            if (optarg[0] == '-') usage(EXIT_FAILURE);
            str = optarg;
            while (1) {
                x = strtol(str, &end, 0);
                if ((end == str) || (x < 1) || (x > 0xff)) {
                    tc_log_error(EXE, "invalid parameter for option -A");
                    exit(1);
                }

                if (*end == '\0') {
                    break;
                }
                if (*end != ',') {
                    tc_log_error(EXE, "invalid parameter for option -A");
                    exit(1);
                }
                str = end + 1;
                new_pass = tc_realloc(pass, (npass + 1) * sizeof (int));
                if (new_pass == NULL) {
                    tc_log_error(EXE, "out of memory");
                    exit(1);
                }
                pass = new_pass;
                pass[npass] = (int)x;
                npass++;
            }
            break;

          case 'M':
            if (optarg[0] == '-') usage(EXIT_FAILURE);
            demux_mode = atoi(optarg);

            if (demux_mode == TC_DEMUX_OFF)
                verbose = TC_QUIET;
            if (demux_mode < 0 || demux_mode > TC_DEMUX_MAX_OPTS) {
                tc_log_error(EXE, "invalid parameter for option -M");
                exit(1);
            }
            break;

          case 'a':
            if (optarg[0] == '-') usage(EXIT_FAILURE);

            if ((n = sscanf(optarg,"%d,%d", &a_track, &v_track)) <= 0) {
                tc_log_error(EXE, "invalid parameter for option -a");
                exit(1);
            }
            break;

          case 'v':
            version();
            exit(0);
            break;

          case 'h':
            usage(EXIT_SUCCESS);
          default:
            usage(EXIT_FAILURE);
        }
    }

    ac_init(AC_ALL);

    /* ------------------------------------------------------------
     * fill out defaults for info structure
     * ------------------------------------------------------------*/

    // assume defaults
    if (name == NULL)
        stream_stype=TC_STYPE_STDIN;

    // no autodetection yet
    if (argc == 1) {
        usage(EXIT_FAILURE);
    }

    // do not try to mess with the stream
    if (stream_stype == TC_STYPE_STDIN) {
        ipipe.fd_in = STDIN_FILENO;
    } else {
        if (tc_file_check(name))
            exit(1);

        ipipe.fd_in = xio_open(name, O_RDONLY);
        if (ipipe.fd_in < 0) {
            tc_log_perror(EXE, "open file");
            exit(1);
        }

        // try to find out the filetype
        stream_magic = fileinfo(ipipe.fd_in, 0);

        if (verbose)
            tc_log_msg(EXE, "(pid=%d) %s", getpid(), filetype(stream_magic));
    }

    // fill out defaults for info structure
    ipipe.fd_out = STDOUT_FILENO;

    ipipe.magic = stream_magic;
    ipipe.stype = stream_stype;
    ipipe.codec = stream_codec;

    ipipe.verbose = verbose;

    ipipe.ps_unit = unit_seek;
    ipipe.ps_seq1 = resync_seq1;
    ipipe.ps_seq2 = resync_seq2;

    ipipe.demux  = demux_mode;
    ipipe.select = pack_sl;
    ipipe.keep_seq = keep_initial_seq;
    ipipe.subid = subid;
    ipipe.fps = fps;

    ipipe.hard_fps_flag = hard_fps_flag;
    ipipe.track = a_track;
    ipipe.name  = logfile;

    //FIXME: video defaults to 0

    /* ------------------------------------------------------------
     * main processing mode
     * ------------------------------------------------------------*/

    if (npass > 0)
        tcdemux_pass_through(&ipipe, pass, npass);
    else
        tcdemux_thread(&ipipe);

    return 0;
}

#include "libtcutil/static_xio.h"

/*************************************************************************/

/*
 * Local variables:
 *   c-file-style: "stroustrup"
 *   c-file-offsets: ((case-label . *) (statement-case-intro . *))
 *   indent-tabs-mode: nil
 * End:
 *
 * vim: expandtab shiftwidth=4:
 */

